toolkit/tools/pkg/srpmpacker/srpmpacker.go

```diff
@@ -0,0 +1,994 @@
+package srpmpacker
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/buildpipeline"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/directory"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/file"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/jsonutils"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/logger"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/network"
+	packagelist "github.com/microsoft/CBL-Mariner/toolkit/tools/internal/packlist"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/retry"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/rpm"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/timestamp"
+)
+
+const (
+	SignatureEnforceString   = "enforce"
+	SignatureSkipCheckString = "skip"
+	SignatureUpdateString    = "update"
+)
+
+type fileSignaturesWrapper struct {
+	FileSignatures map[string]string `json:"Signatures"`
+}
+
+const (
+	srpmOutDir     = "SRPMS"
+	srpmSPECDir    = "SPECS"
+	srpmSOURCESDir = "SOURCES"
+)
+
+type fileType int
+
+const (
+	fileTypePatch  fileType = iota
+	fileTypeSource fileType = iota
+)
+
+type signatureHandlingType int
+
+const (
+	signatureEnforce   signatureHandlingType = iota
+	signatureSkipCheck signatureHandlingType = iota
+	signatureUpdate    signatureHandlingType = iota
+)
+
+// sourceRetrievalConfiguration holds information on where to hydrate files from.
+type sourceRetrievalConfiguration struct {
+	localSourceDir string
+	sourceURL      string
+	caCerts        *x509.CertPool
+	tlsCerts       []tls.Certificate
+
+	signatureHandling signatureHandlingType
+	signatureLookup   map[string]string
+}
+
+// packResult holds the worker results from packing a SPEC file into an SRPM.
+type packResult struct {
+	specFile string
+	srpmFile string
+	err      error
+}
+
+// specState holds the state of a SPEC file: if it should be packed and the resulting SRPM if it is.
+type specState struct {
+	specFile string
+	srpmFile string
+	toPack   bool
+	err      error
+}
+
+func (cfg *Config) CreateAllSrpmsWrapper() error {
+	// Create a template configuration that all packed SRPM will be based on.
+	var templateSrcConfig sourceRetrievalConfiguration
+
+	switch cfg.SignatureHandling {
+	case SignatureEnforceString:
+		templateSrcConfig.signatureHandling = signatureEnforce
+	case SignatureSkipCheckString:
+		logger.Log.Warn("Skipping signature enforcement")
+		templateSrcConfig.signatureHandling = signatureSkipCheck
+	case SignatureUpdateString:
+		logger.Log.Warn("Will update signature files as needed")
+		templateSrcConfig.signatureHandling = signatureUpdate
+	default:
+		logger.Log.Fatalf("Invalid signature handling encountered: %s. Allowed: %s", cfg.SignatureHandling, cfg.ValidSignatureLevels)
```

Comment:
I suspect these panics/fatals only exist because this code was originally under the `main` function. They should probably be converted to normal error handling.

Reply:
Yes I agree, but I think in the scope of the current PR, IMHO we should keep them as is. We should do a treewide refactor in a subsequent PR where we remove all panics and Fatals from any leaf level functions & have a top-level function that does the panic.
Our API design IMO should be to bubble errors up from a function instead of `painc`/`Fatal` in a given function.

What are your thoughts ?

Reply:
Yes, ideally there should only a single panic/fatal call in the main function. Everything else should be normal error handling.

Tree-wide refactors are generally pretty painful to get through PR. So, it might be better to break it up into separate changes for each tool, like you are doing in this PR.

Also, it might be better to fix the error handling first and then do this change where you split-off the code into a separate module. That way, you don’t have an intermediate time where the code is in a bad state.